False positive for empty_xctest_method when assigning statements to _

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

I'm getting a false positive for `empty_xctest_method` with code like the following:

```
    func testStuff() {
        _ = 4 + 4
    }
```

and 

```
    func testStuff() {
        _ = 4 + 4
        _ = 4 + 4
    }
```

But not

```
    func testStuff() {
        let aaa = 4 + 4
    }
```

or 

```
    func testStuff() {
        _ = 4 + 4
        let aaa = 4 + 4
    }
```

When these methods are members of a subclass of XCTestCase

Sample input:

```
import XCTest

class MyTests: XCTestCase {
    func testStuff() {
        _ = 4 + 4
    }
}
```

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint --config swiftlint.yml  t.swift

Linting Swift files at paths t.swift
Linting 't.swift' (1/1)
/tmp/t.swift:4:5: warning: Empty XCTest Method Violation: Empty XCTest method should be avoided. (empty_xctest_method)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

* SwiftLint version - 0.43.1
* Installation method used - HomeBrew
* Paste your configuration file:

```yml
opt_in_rules:
  - empty_xctest_method
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?

No

* Which Xcode version are you using (check `xcodebuild -version`)?

```
Xcode 12.4
Build version 12D4e
```

* Do you have a sample that shows the issue?

```swift
import XCTest

class MyTests: XCTestCase {
    func testStuff() {
        _ = 4 + 4
    }
}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -21,7 +21,7 @@ accordingly._
 * Migrate `empty_xctest_method` rule to SwiftSyntax fixing some false positives.  
   [SimplyDanny](https://github.com/SimplyDanny)
   [#3647](https://github.com/realm/SwiftLint/issues/3647)
-  [#3691](https://github.com/realm/SwiftLint/issues/3691))
+  [#3691](https://github.com/realm/SwiftLint/issues/3691)
 
 ## 0.49.0: Asynchronous Defuzzer
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,7 +18,10 @@ accordingly._
 
 #### Bug Fixes
 
-* None.
+* Migrate `empty_xctest_method` rule to SwiftSyntax fixing some false positives.  
+  [SimplyDanny](https://github.com/SimplyDanny)
+  [#3647](https://github.com/realm/SwiftLint/issues/3647)
+  [#3691](https://github.com/realm/SwiftLint/issues/3691))
 
 ## 0.49.0: Asynchronous Defuzzer
 

--- a/Source/SwiftLintFramework/Rules/Lint/EmptyXCTestMethodRule.swift
+++ b/Source/SwiftLintFramework/Rules/Lint/EmptyXCTestMethodRule.swift
@@ -1,6 +1,6 @@
-import SourceKittenFramework
+import SwiftSyntax
 
-public struct EmptyXCTestMethodRule: Rule, OptInRule, ConfigurationProviderRule {
+public struct EmptyXCTestMethodRule: OptInRule, ConfigurationProviderRule, SwiftSyntaxRule {
     public var configuration = SeverityConfiguration(.warning)
 
     public init() {}
@@ -14,34 +14,52 @@ public struct EmptyXCTestMethodRule: Rule, OptInRule, ConfigurationProviderRule 
         triggeringExamples: EmptyXCTestMethodRuleExamples.triggeringExamples
     )
 
-    public func validate(file: SwiftLintFile) -> [StyleViolation] {
-        return testClasses(in: file).flatMap { violations(in: file, for: $0) }
+    public func makeVisitor(file: SwiftLintFile) -> ViolationsSyntaxVisitor? {
+        EmptyXCTestMethodRuleVisitor()
+    }
+}
+
+private class EmptyXCTestMethodRuleVisitor: SyntaxVisitor, ViolationsSyntaxVisitor {
+    private(set) var violationPositions: [AbsolutePosition] = []
+
+    override func visit(_ node: ClassDeclSyntax) -> SyntaxVisitorContinueKind {
+        guard let inheritanceList = node.inheritanceClause?.inheritedTypeCollection else {
+            return .skipChildren
+        }
+        let isTestClass = inheritanceList
+            .map(\.typeName)
+            .compactMap { $0.as(SimpleTypeIdentifierSyntax.self) }
+            .map(\.name)
+            .map(\.text)
+            .contains("XCTestCase")
+        return isTestClass ? .visitChildren : .skipChildren
     }
 
-    // MARK: - Private
-
-    private func testClasses(in file: SwiftLintFile) -> [SourceKittenDictionary] {
-        let dict = file.structureDictionary
-        return dict.substructure.filter { dictionary in
-            guard dictionary.declarationKind == .class else { return false }
-            return dictionary.inheritedTypes.contains("XCTestCase")
+    override func visitPost(_ node: FunctionDeclSyntax) {
+        if node.isOverride, node.hasEmptyBody {
+            violationPositions.append(node.funcKeyword.positionAfterSkippingLeadingTrivia)
+        } else if node.isTestMethod, node.hasEmptyBody {
+            violationPositions.append(node.funcKeyword.positionAfterSkippingLeadingTrivia)
         }
     }
+}
 
-    private func violations(in file: SwiftLintFile,
-                            for dictionary: SourceKittenDictionary) -> [StyleViolation] {
-        return dictionary.substructure.compactMap { subDictionary -> StyleViolation? in
-            guard
-                let kind = subDictionary.declarationKind,
-                let name = subDictionary.name,
-                XCTestHelpers.isXCTestMember(kind: kind, name: name, dictionary: subDictionary),
-                let offset = subDictionary.offset,
-                subDictionary.enclosedVarParameters.isEmpty,
-                subDictionary.substructure.isEmpty else { return nil }
-
-            return StyleViolation(ruleDescription: Self.description,
-                                  severity: configuration.severity,
-                                  location: Location(file: file, byteOffset: offset))
+private extension FunctionDeclSyntax {
+    var isOverride: Bool {
+        if let modifiers = modifiers {
+            return modifiers.map(\.name).map(\.text).contains("override")
         }
+        return false
+    }
+
+    var hasEmptyBody: Bool {
+        if let body = body {
+            return body.statements.isEmpty
+        }
+        return false
+    }
+
+    var isTestMethod: Bool {
+        identifier.text.hasPrefix("test") && signature.input.parameterList.isEmpty
     }
 }

--- a/Source/SwiftLintFramework/Rules/Lint/EmptyXCTestMethodRule.swift
+++ b/Source/SwiftLintFramework/Rules/Lint/EmptyXCTestMethodRule.swift
@@ -35,6 +35,18 @@ private class EmptyXCTestMethodRuleVisitor: SyntaxVisitor, ViolationsSyntaxVisit
         return isTestClass ? .visitChildren : .skipChildren
     }
 
+    override func visit(_ node: ExtensionDeclSyntax) -> SyntaxVisitorContinueKind {
+        .skipChildren
+    }
+
+    override func visit(_ node: StructDeclSyntax) -> SyntaxVisitorContinueKind {
+        .skipChildren
+    }
+
+    override func visit(_ node: EnumDeclSyntax) -> SyntaxVisitorContinueKind {
+        .skipChildren
+    }
+
     override func visitPost(_ node: FunctionDeclSyntax) {
         if node.isOverride, node.hasEmptyBody {
             violationPositions.append(node.funcKeyword.positionAfterSkippingLeadingTrivia)

--- a/Source/SwiftLintFramework/Rules/Lint/EmptyXCTestMethodRule.swift
+++ b/Source/SwiftLintFramework/Rules/Lint/EmptyXCTestMethodRule.swift
@@ -19,20 +19,11 @@ public struct EmptyXCTestMethodRule: OptInRule, ConfigurationProviderRule, Swift
     }
 }
 
-private class EmptyXCTestMethodRuleVisitor: SyntaxVisitor, ViolationsSyntaxVisitor {
+private final class EmptyXCTestMethodRuleVisitor: SyntaxVisitor, ViolationsSyntaxVisitor {
     private(set) var violationPositions: [AbsolutePosition] = []
 
     override func visit(_ node: ClassDeclSyntax) -> SyntaxVisitorContinueKind {
-        guard let inheritanceList = node.inheritanceClause?.inheritedTypeCollection else {
-            return .skipChildren
-        }
-        let isTestClass = inheritanceList
-            .map(\.typeName)
-            .compactMap { $0.as(SimpleTypeIdentifierSyntax.self) }
-            .map(\.name)
-            .map(\.text)
-            .contains("XCTestCase")
-        return isTestClass ? .visitChildren : .skipChildren
+        node.isXCTestCase ? .visitChildren : .skipChildren
     }
 
     override func visit(_ node: ExtensionDeclSyntax) -> SyntaxVisitorContinueKind {
@@ -48,20 +39,26 @@ private class EmptyXCTestMethodRuleVisitor: SyntaxVisitor, ViolationsSyntaxVisit
     }
 
     override func visitPost(_ node: FunctionDeclSyntax) {
-        if node.isOverride, node.hasEmptyBody {
+        if (node.isOverride || node.isTestMethod) && node.hasEmptyBody {
             violationPositions.append(node.funcKeyword.positionAfterSkippingLeadingTrivia)
-        } else if node.isTestMethod, node.hasEmptyBody {
-            violationPositions.append(node.funcKeyword.positionAfterSkippingLeadingTrivia)
+        }
+    }
+}
+
+private extension ClassDeclSyntax {
+    var isXCTestCase: Bool {
+        guard let inheritanceList = inheritanceClause?.inheritedTypeCollection else {
+            return false
+        }
+        return inheritanceList.contains { type in
+            type.typeName.as(SimpleTypeIdentifierSyntax.self)?.name.text == "XCTestCase"
         }
     }
 }
 
 private extension FunctionDeclSyntax {
     var isOverride: Bool {
-        if let modifiers = modifiers {
-            return modifiers.map(\.name).map(\.text).contains("override")
-        }
-        return false
+        modifiers?.contains { $0.name.text == "override" } ?? false
     }
 
     var hasEmptyBody: Bool {

--- a/Source/SwiftLintFramework/Rules/Lint/EmptyXCTestMethodRuleExamples.swift
+++ b/Source/SwiftLintFramework/Rules/Lint/EmptyXCTestMethodRuleExamples.swift
@@ -84,7 +84,21 @@ internal struct EmptyXCTestMethodRuleExamples {
 
             func testFoo() { XCTAssert(true) }
         }
-        """)
+        """),
+
+        Example("""
+        extension C {
+            override func foo(a: Int) {}
+        }
+
+        struct S {
+            override func foo(a: Int) {}
+        }
+
+        enum E {
+            override func foo(a: Int) {}
+        }
+        """, excludeFromDocumentation: true)
     ]
 
     static let triggeringExamples = [

--- a/Source/SwiftLintFramework/Rules/Lint/EmptyXCTestMethodRuleExamples.swift
+++ b/Source/SwiftLintFramework/Rules/Lint/EmptyXCTestMethodRuleExamples.swift
@@ -11,6 +11,10 @@ internal struct EmptyXCTestMethodRuleExamples {
                 foobar = Foobar()
             }
 
+            override func setUpWithError() throws {
+                foobar = nil
+            }
+
             override func tearDown() {
                 foobar = nil
                 super.tearDown()
@@ -26,6 +30,10 @@ internal struct EmptyXCTestMethodRuleExamples {
                 XCTAssertFalse(foobar?.bar)
 
                 // comment...
+            }
+
+            func testBaz() {
+                _ = 4 + 4
             }
         }
         """),
@@ -63,6 +71,18 @@ internal struct EmptyXCTestMethodRuleExamples {
             func testFoo() { XCTAssertTrue(foobar?.foo) }
 
             func testBar() { XCTAssertFalse(foobar?.bar) }
+        }
+        """),
+
+        // Class/static variables
+
+        Example("""
+        class TotoTests: XCTestCase {
+            override class var runsForEachTargetApplicationUIConfiguration: Bool { true }
+
+            static var allTests = [("testFoo", testFoo)]
+
+            func testFoo() { XCTAssert(true) }
         }
         """)
     ]

